compiler/rustc_middle/src/mir/interpret/allocation/provenance_map.rs

```diff
@@ -364,29 +348,30 @@ impl<Prov: Provenance> ProvenanceMap<Prov> {
             trace!("byte provenances: {bytes:?}");
 
             // And again a buffer for the new list on the target side.
-            let mut dest_bytes = Vec::with_capacity(bytes.len() * (count as usize));
-            for i in 0..count {
-                dest_bytes
-                    .extend(bytes.iter().map(|&(offset, reloc)| (shift_offset(i, offset), reloc)));
-            }
-            debug_assert_eq!(dest_bytes.len(), dest_bytes.capacity());
-            dest_bytes_box = Some(dest_bytes.into_boxed_slice());
+            bytes_box = bytes.into_boxed_slice();
         }
 
-        ProvenanceCopy { dest_ptrs: dest_ptrs_box, dest_bytes: dest_bytes_box }
+        ProvenanceCopy { ptrs: ptrs_box, bytes: bytes_box }
     }
 
     /// Applies a provenance copy.
     /// The affected range, as defined in the parameters to `prepare_copy` is expected
     /// to be clear of provenance.
-    pub fn apply_copy(&mut self, copy: ProvenanceCopy<Prov>) {
-        if let Some(dest_ptrs) = copy.dest_ptrs {
-            self.ptrs.insert_presorted(dest_ptrs.into());
+    pub fn apply_copy(&mut self, copy: ProvenanceCopy<Prov>, range: AllocRange, repeat: u64) {
+        let shift_offset = |idx: u64, offset: Size| offset + range.start + idx * range.size;
+        if !copy.ptrs.is_empty() {
+            for i in 0..repeat {
+                self.ptrs.insert_presorted(
+                    copy.ptrs.iter().map(|&(offset, reloc)| (shift_offset(i, offset), reloc)),
+                );
+            }
```

Comment:
This is the one change that may cost us performance... if there's a bunch of provenance *after* the range we are copying into, but none *in* the given range (i.e. we are hitting the fast-path in insert_presorted), it'll be moved around each time.

We'd need a single iterator yielding the entire repeated range to avoid this... which I guess is possible. Maybe I should try that.

Reply:
it's pretty common for such operations to only be used when the rest is uninit. I guess that will still pessimize the other cases, but without more benchmarks I don't know if we should try anything.